src/type_checker.cpp

```diff
@@ -67,15 +68,18 @@ void TypeChecker::Visit(FuncDefNode& func_def) {
   if (env_.Probe(func_def.id)) {
     // TODO: redefinition of function id
   } else {
+    std::vector<ExprType> param_types{};
+    for (auto& parameter : func_def.parameters) {
+      parameter->Accept(*this);
+      param_types.push_back(parameter->type);
+    }
+
     auto symbol = std::make_unique<SymbolEntry>(func_def.id);
     symbol->expr_type = func_def.return_type;
+    symbol->param_types = param_types;
```

Comment:
Is there any particular reason preventing you from constructing the symbol beforehand and directly pushing the parameter types back into `symbol->param_types` during the visitation of parameters?

Here's the code that expresses my intention:

```cpp
    auto symbol = std::make_unique<SymbolEntry>(func_def.id);
    symbol->expr_type = func_def.return_type;
    for (auto& parameter : func_def.parameters) {
      parameter->Accept(*this);
      symbol->param_types.push_back(parameter->type);
    }
    env_.Add(std::move(symbol));
```

I understand that these approaches differ slightly in error handling. Are we considering this aspect?

Reply:
This one is better!